### ISSUE TITLE
Qwen 3 Embeddings support

### ISSUE THREAD

**Issue by achibb** [UKPLab/sentence-transformers#3383]
Hi there!

Maybe a bit early but just wanted to ask if we can calculate with support of the new Qwen-3 Embeddings models?

(So far only the GGUFs are released, those will probably not work?)

Thanks!

**Comment by tomaarsen** [comment]
Hello!

I'll try and get support for the Qwen3 embedding and reranker models as soon as they're fully released.
Currently, it seems that only https://huggingface.co/Qwen/Qwen3-Embedding-0.6B-GGUF is public. Stay tuned!

- Tom Aarsen

**Comment by tomaarsen** [comment]
They should work with the scripts in these pull requests now:
* https://huggingface.co/Qwen/Qwen3-Embedding-0.6B/discussions/2
* https://huggingface.co/Qwen/Qwen3-Embedding-4B/discussions/1
* https://huggingface.co/Qwen/Qwen3-Embedding-8B/discussions/1

When merged, they'll work in Sentence Transformers like all other embedding models.

- Tom Aarsen

**Comment by nikitaCodeSave** [comment]
I am eagerly awaiting their full integration. Thank you for your product!

**Comment by achibb** [comment]
Great - fast as always - thank you very much! :-) 

Sadly training still seems to give lots of OOM and generally consume alot and take long, even with flash attn2, even with batch size 4 on 24GB plus 1024 max seq len. 

Do you think there will be a way for a more consumer friendly approach?


Still thank you very much for the quick implementation as always!

**Comment by pesuchin** [comment]
@achibb 
Hello. What loss function are you using? If you are using `MultipleNegativesSymmetricRankingLoss`, you might be able to reduce GPU memory consumption by using `CachedMultipleNegativesSymmetricRankingLoss`, setting the mini-batch size to 1, and using a larger batch size.

**Comment by tomaarsen** [comment]
@pesuchin 's recommendation is solid. Beyond that, there's not much that can be done I think. Be sure to use bfloat16 or float16 (`bf16=True` or `fp16=True` in the training arguments)

- Tom Aarsen

**Comment by BBC-Esq** [comment]
I've tested the new qwen3 embedding models and it's working in my program.  Okay to close?

**Comment by claeyz** [comment]
@BBC-Esq I'm trying to get it working and still currently facing several issues. Could you tell me which versions of 
- transformers
- sentence-transformers
- pytorch

are you using ?

Thanks

**Comment by BBC-Esq** [comment]
@claeyz If I recall correctly I was using:

transformers 4.52.4
torch 2.6.0+cu124
sentence-transformers 4.1.0

**Comment by achibb** [comment]
@tomaarsen - btw: I figured that the memory issues come mostly from the context length.

basically adding max_length=512 (or whatever you prefer) to it will make it work just like anything else. Should we consider adding it / a lower batch size or a comment or something by default?

**Comment by tomaarsen** [comment]
Fair enough, lots of models with a long sequence length work with high batch sizes with small texts, but require much more memory for larger texts. The choice for the maximum sequence length is on the model author, Qwen in this case, so I don't think we should lower the batch size by default. Perhaps a comment would be useful on the model repositories? 

E.g. like updating the script from https://huggingface.co/Qwen/Qwen3-Embedding-0.6B
```python
# Requires transformers>=4.51.0
# Requires sentence-transformers>=2.7.0

from sentence_transformers import SentenceTransformer

# Load the model
model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")

# We recommend enabling flash_attention_2 for better acceleration and memory saving,
# together with setting `padding_side` to "left":
# model = SentenceTransformer(
#     "Qwen/Qwen3-Embedding-0.6B",
#     model_kwargs={"attn_implementation": "flash_attention_2", "device_map": "auto"},
#     tokenizer_kwargs={"padding_side": "left"},
# )
# Optionally, lower the maximum sequence length for lower memory usage
# model.max_seq_length = 512

# The queries and documents to embed
queries = [
    "What is the capital of China?",
    "Explain gravity",
]
documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun.",
]

# Encode the queries and documents. Note that queries benefit from using a prompt
# Here we use the prompt called "query" stored under `model.prompts`, but you can
# also pass your own prompt via the `prompt` argument
query_embeddings = model.encode(queries, prompt_name="query")
document_embeddings = model.encode(documents)

# Compute the (cosine) similarity between the query and document embeddings
similarity = model.similarity(query_embeddings, document_embeddings)
print(similarity)
# tensor([[0.7646, 0.1414],
#         [0.1355, 0.6000]])
```
But I'm not sure if the model authors are receptive to such PRs.

- Tom Aarsen

**Comment by achibb** [comment]
Sounds good, yes. If you want me to I can do it, or you, whatever you prefer :).

**Comment by tomaarsen** [comment]
Feel free to make a PR if you wish!
